sound/soc/sof/intel/hda-dai.c

```diff
@@ -161,11 +163,23 @@ static int hda_link_pcm_trigger(struct snd_pcm_substream *substream,
 		/* fallthrough */
 	case SNDRV_PCM_TRIGGER_START:
 	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
+		/*
+		 * Hostless streams do not use host DMA but the host DMA
+		 * channel with the same ID as the link DMA must be
+		 * reserved to prevent it from being allocated to another
+		 * stream.
+		 */
+		if (!stream.opened)
+			stream.opened = 1;
 		snd_hdac_ext_link_stream_start(link_dev);
 		break;
-	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
 	case SNDRV_PCM_TRIGGER_SUSPEND:
 	case SNDRV_PCM_TRIGGER_STOP:
+		if (stream.opened)
+			stream.opened = 0;
```

Comment:
We directly handle internals of the hdac library but those assignments are only taken into account when there is an explicit call to hdac_ext_xyz(), so it's quite hard to follow the state machine here.

Reply:
@plbossart understood. I probably need to add some more comments to why and how this is done here. Will address in v2